Sources/SmokeHTTP1/SmokeHTTP1Server.swift

```diff
@@ -89,20 +108,56 @@ public class SmokeHTTP1Server {
             .childChannelOption(ChannelOptions.allowRemoteHalfClosure, value: true)
         
         channel = try bootstrap.bind(host: ServerDefaults.defaultHost, port: port).wait()
+        
+        fullyShutdownPromise.futureResult.whenComplete { [unowned self] in
+            do {
+                try self.group.syncShutdownGracefully()
+            } catch {
+                Log.error("Server unable to shutdown cleanly following full shutdown.")
+            }
+            
+            Log.info("SmokeHTTP1Server shutdown.")
+        }
+        
+        Log.info("SmokeHTTP1Server started on port \(port).")
     }
     
     /**
-     Stops the server.
+     Initiates the process of shutting down the server.
      */
-    public func stop() throws {
-        try group.syncShutdownGracefully()
-        try threadPool.syncShutdownGracefully()
+    public func shutdown() throws {
+        quiesce.initiateShutdown(promise: fullyShutdownPromise)
     }
     
     /**
      Blocks until the server has been shut down.
      */
-    public func wait() throws {
-        try channel?.closeFuture.wait()
+    public func waitUntilShutdown() throws {
+        try fullyShutdownPromise.futureResult.wait()
+    }
+    
+    /**
+     Blocks until the server has been shut down. After the server
+     has been fully shutdown, the provided closure will be executed.
+     
+     - Parameters:
+        - onShutdown: the closure to be executed after the server has been
+                      fully shutdown.
+     */
+    public func waitUntilShutdownAndThen(onShutdown: @escaping () -> Void) throws {
+        fullyShutdownPromise.futureResult.whenComplete(onShutdown)
```

Comment:
@tachyonics just as a note: Attaching promises to this particular future is a bit difficult. The reason is that further up, you shut the `EventLoopGroup` for the `EventLoop` that this future is bound to down. But a promise/future whose `EventLoop` is shut down is pretty dangerous as it can't be used anymore because it needs its `EventLoop` do to anything useful so that can easily lead to a deadlock.

As the patch stands today I suspect this `waitUntilShutdownAndThen` method will only work if called _before_ `fullyShutdownPromise` has been fulfilled which isn't a great API. 

If I were you I would probabaly
- make sure that `initiateShutdown` is only ever called once, no matter what the user does
- (this is annoying but I think necessary) store some private state (needs synchronisation with a lock/dispatchqueue), for example
```swift
struct LifecycleState {
    enum State {
        case upAndRunning
        case shuttingDown
        case shutDown
    }
    var shutdownCompletionHandlers: [() -> Void] = []
    var shutdownState: State = .upAndRunning
}
```

Then you can do the right stuff:
- in `shutdown` you shift the state from `.upAndRunning` to `.shuttingDown`, if it's not `.upAndRunning`, then you throw an error, or ignore, or crash, whatever you like best
- in `onShutdown` and `waitUntilShutdown` you briefly take the lock and check: if `.upAndRunning` or `.shuttingDown` you'll need to add the completion handler to the `shutdownCompletionHandlers` array. If it's already `.shutDown` you just invoke the completion handler straight away
- when the quiescing has completed, you shift the state from `.shuttingDown` to `.shutDown` and you run (outside the lock) all the completion handlers in the `shutdownCompletionHandlers` array and set the array to the empty array.

That way, the following three things should work

1. waiting for a shutdown before it has been initiated (which already works just fine with your proposed patch)
2. waiting for a shutdown whilst the shutdown is in progress (which sometimes (most often I'd think) already works with your proposed patch)
3. waiting for a shutdown after it has completed (which does not work with your proposed patch)

Does that make sense? The reason this is so hard is because here we bridge from a world with `EventLoopFuture/Promise`s and `EventLoop`s to one without. So we're basically shutting down the execution mechanism so we can't use it anymore. `Dispatch` just can't be shut down, hence no such issues.

Reply:
the other option obviously is doing away with being able to be notified (`onShutdown`) about or  waiting (`waitUntilShutdownAndThen`) for the shutdown. If you just had one method `syncShutdown` which initiates the shutdown and only returns when complete, it'd be a bit simpler but you'd still need to keep track of the state (user could call it twice / the signal source is also initiating shutdowns and that might race against the user).

Reply:
@weissi thanks for all your recommendations. I have proposed changes that hopefully follow your comments. Let me know if there is anything I have missed.